index.rst

```diff
@@ -1,57 +1,236 @@
+A blocking-with-callback or other asynchronous API to the Butler could reduce the second source of latency, but it would not affect the first.
+
+Copying the image from its publication point to the handler-local repo is assumed to be an irreducible cost.
+By placing the image in permanent storage as soon as possible after retrieval from the Camera Data Acquisition system, the possibility of loss is minimized.
+Previously, Prompt Processing was to receive a special, low-latency copy of the image (with firmware-based crosstalk correction), while the permanent archive would receive a non-crosstalk-corrected version after some delay (after the network burst caused by Prompt Processing).
+Since the Camera is no longer doing firmware crosstalk correction, moving the copy to the destination end of the international link avoids congesting that link, places it in a high-bandwidth local network environment, and enables all potential USDF users of the images to receive the same low-latency service.
+
+The Butler interactions are handled via the Python API for export and import, and, for simplicity of invocation, the command line tooling for ingest and pipeline execution.
+
+Using a batch queue mechanism for firing off the pipeline execution is attractive in that the HTTP handler would not need to persist for the lifetime of the pipeline, and resilience and elasticity could be built into the batch system.
+On the other hand, it poses serious issues with regard to latency and/or communications.
+If the batch system were to be triggered where ``pipetask`` is invoked in the current design, the latency of batch submission, queueing, dequeueing, and execution startup would be in the critical path.
+In addition, it would be difficult to preload data for the batch job, as the worker would not be known until batch submission.
+If the batch system were instead to be triggered immediately by the POST handler, latency would not be an issue.
+Instead, it would be difficult for the batch job to be notified of image arrival.
+Either a Butler poll (undesirable for reasons given above) or special messaging subscription code in the batch job would be necessary.
+Neither of these seems to offer much of an advantage over the simpler design of including the preload and pipeline execution in the same handler process, as long as that handler is resilient and elastically scalable, which it is.
+
+One limitation of the design is that the pipeline to be executed generally must use data from a single detector/CCD only.
+The handler-local repo only has data for that detector, and there is no way to access data in other handlers' repos.
+An out-of-band, non-Butler mechanism could be invented to communicate gather/scatter information, but this would add substantial complexity and new code.
+If such pipelines were to be necessary, a shared repo would likely be needed.
```

Comment:
We discussed this morning that this is _probably_ a safe assumption for LSST instruments, but if we started developing a per-detector architecture and then discovered that we do need cross-detector processing, how hard would it be to change things?

Reply:
The simplest change that would still be reliable would be to get rid of the handler-local preload Butler repo and go with a design that waits for all images to arrive followed by a BPS submission to a batch system using a shared read/write Butler repo.  This makes AP execution the same as DRP.  Of course the problem with this is that latency goes up by at least 20 seconds, in addition to needing a high-performance shared repo.